cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -781,4 +799,34 @@ private void writeSystemMetadata(Id.Artifact artifactId, ArtifactInfo artifactIn
     ArtifactSystemMetadataWriter writer = new ArtifactSystemMetadataWriter(metadataStore, artifactId, artifactInfo);
     writer.write();
   }
+
+  public String callPluginMethod(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                 String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                 String methodName, String requestBody)
+    throws NotFoundException, IOException, InvalidArtifactException {
+
+    try (CloseableClassLoader parentClassLoader =
+           artifactClassLoaderFactory.createClassLoader(parentArtifactDescriptor.getLocation())) {
+      Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+      File stageDir = DirUtils.createTempDir(baseUnpackDir);
+      try (PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir)) {
+        // we pass the parent artifact to endpoint plugin context, as the endpoint can try to load a plugin
+        // and we provide app as the parent for plugins.
+        // using parent artifact will help in finding the plugins from endpoint method.
+        DefaultEndpointPluginContext defaultEndpointPluginContext =
+          new DefaultEndpointPluginContext(namespace, this, pluginInstantiator,
+                                           Id.Artifact.from(namespace, parentArtifactDescriptor.getArtifactId()));
+
+        pluginInstantiator.addArtifact(artifactDetail.getDescriptor().getLocation(), artifactId.toArtifactId());
+        File artifactFile = new File(artifactDetail.getDescriptor().getLocation().toURI());
```

Comment:
Is it always local? Isn't that artifacts are stored in HDFS in distributed mode?